File: frontend/src/services/api/cc-model-preset.ts
import axios from './axios'

export interface CCModelPresetInfo {
  id: number
  name: string
  description: string
  base_url: string
  auth_token: string
  api_timeout_ms: string
  model_type: 'preset' | 'manual'
  preset_model: string
  anthropic_model: string
  small_fast_model: string
  default_sonnet: string
  default_opus: string
  default_haiku: string
  extra_env: Record<string, string>
  is_default: boolean
  create_time: string
  update_time: string
  config_json: Record<string, unknown>
}

export interface CCModelPresetCreate {
  name: string
  description?: string
  base_url?: string
  auth_token?: string
  api_timeout_ms?: string
  model_type?: 'preset' | 'manual'
  preset_model?: string
  anthropic_model?: string
  small_fast_model?: string
  default_sonnet?: string
  default_opus?: string
  default_haiku?: string
  extra_env?: Record<string, string>
}

export type CCModelPresetUpdate = Partial<CCModelPresetCreate>

export interface CCModelPresetTestItem {
  preset_id: number
  preset_name: string
  model_name: string
  success: boolean
  latency_ms: number
  used_model?: string | null
  response_text?: string | null
  input_tokens?: number
  output_tokens?: number
  error_message?: string | null
}

export const ccModelPresetApi = {
  getList: async (): Promise<CCModelPresetInfo[]> => {
    const res = await axios.get<{ total: number; items: CCModelPresetInfo[] }>('/cc-model-presets/list')
    return res.data.items
  },
  create: async (body: CCModelPresetCreate): Promise<CCModelPresetInfo> => {
    const res = await axios.post<CCModelPresetInfo>('/cc-model-presets', body)
    return res.data
  },
  update: async (id: number, body: CCModelPresetUpdate): Promise<CCModelPresetInfo> => {
    const res = await axios.patch<CCModelPresetInfo>(`/cc-model-presets/${id}`, body)
    return res.data
  },
  delete: async (id: number): Promise<void> => {
    await axios.delete(`/cc-model-presets/${id}`)
  },
  test: async (presetIds: number[]): Promise<CCModelPresetTestItem[]> => {
    const res = await axios.post<{ items: CCModelPresetTestItem[] }>('/cc-model-presets/test', {
      preset_ids: presetIds,
    })
    return res.data.items
  },
}

File: nekro_agent/schemas/cc_model_preset.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class CCModelPresetCreate(BaseModel):
    name: str = Field(..., max_length=64)
    description: str = Field(default="", max_length=512)
    base_url: str = Field(default="https://api.nekro.ai/claude", max_length=256)
    auth_token: str = Field(default="", max_length=256)
    api_timeout_ms: str = Field(default="3000000", max_length=16)
    model_type: Literal["preset", "manual"] = "preset"
    preset_model: str = Field(default="opus", max_length=64)
    anthropic_model: str = Field(default="", max_length=128)
    small_fast_model: str = Field(default="", max_length=128)
    default_sonnet: str = Field(default="", max_length=128)
    default_opus: str = Field(default="", max_length=128)
    default_haiku: str = Field(default="", max_length=128)
    extra_env: Dict[str, str] = {}


class CCModelPresetUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=64)
    description: Optional[str] = Field(default=None, max_length=512)
    base_url: Optional[str] = Field(default=None, max_length=256)
    auth_token: Optional[str] = Field(default=None, max_length=256)
    api_timeout_ms: Optional[str] = Field(default=None, max_length=16)
    model_type: Optional[Literal["preset", "manual"]] = None
    preset_model: Optional[str] = Field(default=None, max_length=64)
    anthropic_model: Optional[str] = Field(default=None, max_length=128)
    small_fast_model: Optional[str] = Field(default=None, max_length=128)
    default_sonnet: Optional[str] = Field(default=None, max_length=128)
    default_opus: Optional[str] = Field(default=None, max_length=128)
    default_haiku: Optional[str] = Field(default=None, max_length=128)
    extra_env: Optional[Dict[str, str]] = None


class CCModelPresetInfo(BaseModel):
    id: int
    name: str
    description: str
    base_url: str
    auth_token: str
    api_timeout_ms: str
    model_type: str
    preset_model: str
    anthropic_model: str
    small_fast_model: str
    default_sonnet: str
    default_opus: str
    default_haiku: str
    extra_env: Dict[str, str]
    is_default: bool
    create_time: str
    update_time: str
    config_json: Dict[str, Any]


class CCModelPresetListResponse(BaseModel):
    total: int
    items: List[CCModelPresetInfo]
